src/eval.rs

```diff
@@ -1,26 +1,998 @@
-//! The interpreter
-
+use crate::parse::parse;
 use crate::syntax::*;
 use std::collections::HashMap as Map;
+use std::ops::Range;
+use std::result;
+use thiserror::Error;
+
+/// Errors that can occur during type checking
+#[derive(Error, Debug)]
 
 enum RuntimeError {
-    InvalidInput,
+    #[error("Invalid expression")]
+    InvalidExpression,
+    #[error("Invalid statement")]
+    InvalidStatement,
+    #[error("Division by zero")]
+    DivisionbyZero,
+    #[error("Type error")]
+    TypeError,
+    #[error("Invalid pperand")]
+    InvalidOperand,
+    #[error("Modulus by zero")]
+    ModulusByZero,
+    #[error("Cast fail: out of expected range")]
+    OutOfRange,
+    #[error("Call fail: incorrect args")]
+    IncorrectArgs,
+}
+
+// abstraction for values making up expressions
+#[derive(Debug, Clone, PartialEq)]
+enum Value {
+    Bool(bool),
+    Num(i64, Range<i64>),
+    Sym(Symbol),
+}
+
+// map of variable names to values
+type Env = Map<Id, Value>;
+
+fn init_env(program: &Program) -> Env {
+    let mut env = Env::new();
+
+    // insert alphabet into env
+    for symbol in &program.alphabet {
+        let id = id(&symbol.to_string());
+        let value = Value::Sym(*symbol);
+        println!("Inserting alpha. id -> {}, value -> {:?}", id, value);
+        env.insert(id, value);
+    }
+
+    // insert locals into env
+    for (id, typ) in &program.locals {
+        let value = match typ {
+            // defaults for now are zero values
+            Type::BoolT => Value::Bool(false),
+
+            //  default to be beginning of range
+            Type::NumT(range) => Value::Num(range.start, range.clone()),
+
+            //  default to empty char
+            Type::SymT => Value::Sym(Symbol(' ')),
+
+            _ => continue, // for now continue, raise error later
+        };
+
+        println!("Inserting local id -> {}, value -> {:?}", id, value);
+        env.insert(id.clone(), value);
+    }
+
+    env
 }
 
-fn eval(program: &Program, input: &str) -> Result<bool, RuntimeError> {
+fn eval(program: &Program, input: &str) -> Result<(bool, Env), RuntimeError> {
     // create initial state
+    let mut env = init_env(program);
 
+    // TODO - figure out what to do with input ("consume each symbol???")
     // consume each symbol
 
     // evaluate the final condition
-    Ok(false)
-}
+    for stmt in &program.action.1 {
+        eval_stmt(stmt, &mut env, &program)?;
+    }
+    eval_expr(&program.accept, &env, &program)?;
 
-// define the state type
-enum Value {}
+    Ok((false, env))
```

Comment:
rather than returning false, you should return the value of the last expression you evaluated (and return a type error if it is not a bool)